File: DS/Stacks/Python/balanced_parentheses.py
def balanced_parentheses(str1):
    stack = []

    for i in str1:
        if i == '(' or i == '[' or i == '{':
            stack.append(i)

        elif i == ')' or i == ']' or i == '}':
            if len(stack) == 0:
                return "parentheses are not balanced"

            elif i == ')' and stack.pop() != '(':
                return "parentheses are not balanced"

            elif i == '}' and stack.pop() != '{':
                return "parentheses are not balanced"

            elif i == ']' and stack.pop() != '[':
                return "parentheses are not balanced"

    if not stack:
        return "parentheses are balanced"

    else:
        return "parentheses are not balanced"


s = "{}[](){[()][]}"
print(balanced_parentheses(s))

output: "parentheses are balanced"

File: DS/Stacks/cpp/balanced_parentheses.cpp
#include <bits/stdc++.h>
using namespace std;
// balanced parentheses = [()()], {[()]}
// not balanced parentheses = )( , [(]), ([){}]

string parentheses(string str1, int n){
    stack<char> S;
    int i;
    for(i=0;i<n;i++){
        if(str1[i]=='(' || str1[i]=='{' || str1[i]=='['){
            S.push(str1[i]);
        }

        else if(str1[i]==')' || str1[i]=='}' || str1[i]==']'){
            if(S.empty()){
                return "parentheses are not balanced";
            }
            else if(str1[i]==')'&& S.top()!='('){
                return "parentheses are not balanced";
            }
            else if(str1[i]=='{'&& S.top()!='{'){
                return "parentheses are not balanced";
            }
            else if(str1[i]=='['&& S.top()!='['){
                return "parentheses are not balanced";
            }
            else S.pop();
        }
    }
    
    return S.empty() ? "parentheses are balanced":"parentheses are not balanced";
}


int main(){
    string str1= "[{()()}]";
    cout<<parentheses(str1, str1.length());
}

// output: parentheses are balanced